From a35f45d887635bd04b626b8733f1449f0acd1109 Mon Sep 17 00:00:00 2001
From: Dave Stevenson <dave.stevenson@raspberrypi.org>
Date: Mon, 29 Oct 2018 14:21:04 +0000
Subject: [PATCH] staging: bcm2835-camera: Ensure H264 header bytes get
 a sensible timestamp

H264 header come from VC with 0 timestamps, which means they get a
strange timestamp when processed with VC/kernel start times,
particularly if used with the inline header option.
Remember the last frame timestamp and use that if set, or otherwise
use the kernel start time.

https://github.com/raspberrypi/linux/issues/1836

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>
---
 .../bcm2835-camera/bcm2835-camera.c           | 28 +++++++++++++++++--
 .../bcm2835-camera/bcm2835-camera.h           |  2 ++
 2 files changed, 28 insertions(+), 2 deletions(-)

--- a/drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.c
+++ b/drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.c
@@ -364,7 +364,12 @@ static void buffer_cb(struct vchiq_mmal_
 		return;
 	}
 
-	if (dev->capture.vc_start_timestamp != -1 && pts) {
+	if (dev->capture.vc_start_timestamp == -1) {
+		buf->vb.vb2_buf.timestamp = ktime_get_ns();
+		v4l2_dbg(1, bcm2835_v4l2_debug, &dev->v4l2_dev,
+			"Buffer time set as current time - %lld",
+			buf->vb.vb2_buf.timestamp);
+	} else if (mmal_buf->pts != 0) {
 		ktime_t timestamp;
 		s64 runtime_us = pts -
 		    dev->capture.vc_start_timestamp;
@@ -377,8 +382,23 @@ static void buffer_cb(struct vchiq_mmal_
 			 ktime_to_ns(timestamp));
 		buf->vb.vb2_buf.timestamp = ktime_to_ns(timestamp);
 	} else {
-		buf->vb.vb2_buf.timestamp = ktime_get_ns();
+		if (dev->capture.last_timestamp) {
+			buf->vb.vb2_buf.timestamp =
+				dev->capture.last_timestamp;
+			v4l2_dbg(1, bcm2835_v4l2_debug,
+				&dev->v4l2_dev,
+				"Buffer time set as last timestamp - %lld",
+				buf->vb.vb2_buf.timestamp);
+		} else {
+			buf->vb.vb2_buf.timestamp =
+				ktime_to_ns(dev->capture.kernel_start_ts);
+			v4l2_dbg(1, bcm2835_v4l2_debug,
+				&dev->v4l2_dev,
+				"Buffer time set as start timestamp - %lld",
+				buf->vb.vb2_buf.timestamp);
+		}
 	}
+	dev->capture.last_timestamp = buf->vb.vb2_buf.timestamp;
 	buf->vb.sequence = dev->capture.sequence++;
 	buf->vb.field = V4L2_FIELD_NONE;
 
@@ -386,6 +406,9 @@ static void buffer_cb(struct vchiq_mmal_
 	if (mmal_flags & MMAL_BUFFER_HEADER_FLAG_KEYFRAME)
 		buf->vb.flags |= V4L2_BUF_FLAG_KEYFRAME;
 
+	v4l2_dbg(1, bcm2835_v4l2_debug, &dev->v4l2_dev,
+		"Buffer has ts %llu",
+		dev->capture.last_timestamp);
 	vb2_buffer_done(&buf->vb.vb2_buf, VB2_BUF_STATE_DONE);
 
 	if (mmal_flags & MMAL_BUFFER_HEADER_FLAG_EOS &&
@@ -549,6 +572,7 @@ static int start_streaming(struct vb2_qu
 	}
 
 	dev->capture.kernel_start_ts = ktime_get();
+	dev->capture.last_timestamp = 0;
 
 	/* enable the camera port */
 	dev->capture.port->cb_ctx = dev;
--- a/drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.h
+++ b/drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.h
@@ -93,6 +93,8 @@ struct bm2835_mmal_dev {
 		ktime_t kernel_start_ts;
 		/* Sequence number of last buffer */
 		u32		sequence;
+		/* Timestamp of last frame */
+		u64		last_timestamp;
 
 		struct vchiq_mmal_port  *port; /* port being used for capture */
 		/* camera port being used for capture */
